### PR TITLE
Zoom capability (mouse wheel) using bigshot

### PR DESCRIPTION
**Pull request by davidrapin**
In the slideshow view, scrolling the mouse wheel will now allow users to zoom in and out of the current image using google-maps-like interactions. Hitting the "0" key on he keyboard will reset the zoom factor to fit the screen.
BigShot project page: https://code.google.com/p/bigshot/

### CODE REVIEW

**Comment by karlitschek** [comment]
@icewind1991 What do you think?

File: appinfo/app.php [hunk 1]
```diff
@@ -35,6 +35,7 @@
 
 // make slideshow available in files and public shares
 OCP\Util::addScript('gallery', 'jquery.mousewheel-3.1.1');
+OCP\Util::addScript('gallery', 'bigshot-compressed');
```
Comment (DeepDiver1975) [inline appinfo/app.php]: please add the uncompressed js and remove the compressed - THX

**Comment by DeepDiver1975** [comment]
nice - I like this! THX

**Comment by libasys** [comment]
this extension is cool, but perfect if we have the option for fullscreen viewing images! this option is given in the example by bigshot demo!

**Comment by libasys** [comment]
added bigshot fullscreen feature:
Demo: https://ssl.webpack.de/demo70.libasyscloud.de/owncloud/index.php/apps/gallery/public/1c17af51f8e6e9d3039faba4106d8fdb

I will do a pull request for interested persons are willing to use the fullscreen feature  if the bigshot feature PR is merged!

Somebody interested?

**Comment by davidrapin** [comment]
@libasys I can add the fullscreen feature in the current pull request if you want !

**Comment by libasys** [comment]
ok perfect do you need some help, or code snippet?

**Comment by davidrapin** [comment]
@libasys I already had a prototype and have taken a look at your code, I'll add your 'fullscreen' icon (which I think is a nice touch) and commit as soon as I have the time! Thanks :)

**Comment by libasys** [comment]
ok perfect!

**Comment by libasys** [comment]
@davidrapin Please add a check for mobile support, cause if you use bigshot with an tablet the mouseevents not working so you can't navigate etc, so we have to deactivate the bigshot feature for mobile.
a Codesnippet to check if the device is mobile for use in the init function of the slideshow:

```
var browser = new bigshot.Browser ();
           if(!browser.browserIsViewporting()){
               this.isMobile=false;
           }else{
             this.isMobile=true;
           }
```

Thanks

**Comment by davidrapin** [comment]
@libasys It took me some time but I managed to make the mobile controls work together with BigShot.js :) Enjoy !

**Comment by libasys** [comment]
I test it today!

**Comment by icewind1991** [comment]
Using `0` to reset the zoom doesn't seem to work when in fullscreen mode (Chromium 37, Linux)

**Comment by davidrapin** [comment]
@icewind1991 Nice catch, I have the same bug with Google Chrome in Windows, so I presume it's a webkit bug. I'll take a look at it after work :smile:

**Comment by libasys** [comment]
Added some more cool Feature outgoing of this PR

# New Features:

I. Added a Preview Navigation of Thumbnails on the bottom (using the cached Thumbnail)
II. Added 3 ZoomModi (Zoom Fit Height, Zoom Fit Width, Zoom Fit Viewport)
III. Choosen ZoomMode active on next or prev Image
IIII. Added function for up/down arrow key (up = show always all controls, down = hide active controls)
IIII. Added swipe left/right for the preview Navigation of Thumbnails on mobiles

I will do a PR after this PR is commited, if somebody is interested in this!

Tested with Mac: Safari, Firefox, Chrome, Opera!

Link with this features:
https://ssl.webpack.de/demo70.libasyscloud.de/owncloud/index.php/apps/gallery/public/1c17af51f8e6e9d3039faba4106d8fdb

Have fun

**Comment by davidrapin** [comment]
@icewind1991 Fixed. Was due to BigShot.js using the [wrong method for the webkit fullscreen API](http://stackoverflow.com/questions/15898459/keyboard-keys-not-working-in-chrome-app-when-in-full-screen). One lowercase 's' and everything in fine again :sunglasses: 
Works nice as tested by me on iPhone, Firefox (windows) and Chrome (windows).

**Comment by icewind1991** [comment]
> I. Added a Preview Navigation of Thumbnails on the bottom (using the cached Thumbnail)
> II. Added 3 ZoomModi (Zoom Fit Height, Zoom Fit Width, Zoom Fit Viewport)

Not sure if I like all the clutter it adds (cc @jancborchardt)

**Comment by libasys** [comment]
@icewind1991 "clutter" ok ...! You are the maintainer of the gallery app. So I don't do a PR to merge all the "clutter".

**Comment by jancborchardt** [comment]
Ok folks, let’s do small steps here.

Adding zooming functionality is great, and we should keep this pull request to that for now.
That is: No fullscreen functionality, no thumbnail preview navigation, no bunch of zoom modes.

@davidrapin @libasys the general functionality looks great! So it would be cool if only the zooming in could be possible. No other functionality or interface elements should be added.
We don’t need zooming out because the image fits the screen automatically. And we need to test on mobile (pinch/spread to zoom).

**Comment by davidrapin** [comment]
@jancborchardt I can remove the fullscreen icon, but is it necessary to remove the fullscreen feature? it adds no complexity to the code nor the interface if we keep it as a keyboard shortcut.

By the way, I have tested the pinch/spread on mobile (Mobile Safari on iPhone): it works. If someone could test on Android (mobile Firefox and Chrome) that would be nice.

**Comment by jancborchardt** [comment]
@davidrapin what is the keyboard shortcut for it?

**Comment by jancborchardt** [comment]
(Asking because we need to make sure it’s standard and same as on other platforms.)

**Comment by davidrapin** [comment]
@jancborchardt The shortcut to toggle fullscreen is "f" (like in VLC).

**Comment by jancborchardt** [comment]
Ok, sounds good. cc @owncloud/designers

**Comment by davidrapin** [comment]
@jancborchardt @icewind1991 As requested, I removed the fullscreen icon.

Summary of the changes for the gallery slideshow:
- new scroll-to-zoom capability (pinch/spread in mobile)
- new fullscreen capability
- new keyboard shortcut: "0" to fit image to screen
- new keyboard shortcut: "f" to toggle fullscreen mode

Tested on Safary Mobile (iPhone), Firefox (Windows) and Chrome (Windows).
Do I need to change anything else before this can be merged?

File: js/slideshow.js [hunk 2]
```diff
@@ -21,11 +21,30 @@ var SlideShow = function (container, images, interval, maxScale) {
...
+	var e = this.container.get(0);
+	this.canFullScreen = e.requestFullscreen !== undefined
+		|| e.mozRequestFullScreen !== undefined
+		|| e.webkitRequestFullscreen !== undefined
+		|| e.msRequestFullscreen !== undefined;
+	
+	// makes UI controls work in mobile version
+	var browser = new bigshot.Browser();
+	this.container.children('input').each(function(i, e) {
+		browser.registerListener(e, "click", browser.stopEventBubblingHandler(), false);
```
Comment (MorrisJobke) [inline js/slideshow.js]: Can I ask you to use `'` instead of `"`? Thanks

File: js/slideshow.js [hunk 3]
```diff
@@ -85,6 +94,36 @@ SlideShow.prototype.init = function (play) {
...
 };
 
+SlideShow.prototype.zoomToFit = function() {
+	if (this.zoomable !== null) {
+		this.zoomable.flyZoomToFit();
+	}
+};
+
+SlideShow.prototype.fullScreenStart = function () {
+	if (!this.canFullScreen) return;
```
Comment (MorrisJobke) [inline js/slideshow.js]: Please always use curly brackets

File: js/slideshow.js [hunk 4]
```diff
@@ -85,6 +94,36 @@ SlideShow.prototype.init = function (play) {
...
+	this.fullScreen = new bigshot.FullScreen(this.container.get(0));
+	this.fullScreen.open();
+	this.fullScreen.addOnClose(function(evt){
+		this.fullScreenExit();
+	}.bind(this));
+};
+
+SlideShow.prototype.fullScreenExit = function () {
+	if (this.fullScreen === null) return;
```
Comment (MorrisJobke) [inline js/slideshow.js]: Same here

File: js/slideshow.js [hunk 5]
```diff
@@ -85,6 +94,36 @@ SlideShow.prototype.init = function (play) {
...
+	this.fullScreen.close();
+	this.fullScreen = null;
+};
+
+SlideShow.prototype.fullScreenToggle = function() {
+	if (this.zoomable === null) return;
```
Comment (MorrisJobke) [inline js/slideshow.js]: And here

File: js/slideshow.js [hunk 6]
```diff
@@ -99,9 +138,28 @@ SlideShow.prototype.show = function (index) {
...
+				container: this.container.get(0),
+				maxZoom: 2,
+				minZoom: -5,
+				touchUI: false,
+				width: image.width,
+				height: image.height
+			}), image);
+			
+			// prevent zoom-on-doubleClick
+			this.zoomable.addEventListener("dblclick", function(ie) {
```
Comment (MorrisJobke) [inline js/slideshow.js]: Please single quotes instead of double quotes.

File: templates/slideshow.html [hunk 7]
```diff
@@ -4,6 +4,6 @@
 	<input type="button" class="svg pause icon-view-pause"/>
 	<input type="button" class="svg previous icon-view-previous"/>
 	<input type="button" class="svg exit icon-view-close"/>
-
+
```
Comment (MorrisJobke) [inline templates/slideshow.html]: Useless whitespace ;)

**Comment by MorrisJobke** [comment]
Can you also squash your changes into one commit after you've done the above fixes?

After that I will test this. Thanks for your work :)

**Comment by davidrapin** [comment]
@MorrisJobke You're welcome, thank you for the feedback :smile: I fixed the style issues, let me know if anything is missing.

**Comment by jancborchardt** [comment]
@davidrapin did you also disable the possibility of zooming in? We do not need zooming out (as in, making the image smaller) as it’s already fit to the page.

**Comment by davidrapin** [comment]
@jancborchardt Zooming out so as to make the image smaller than the screen is no longer possible.

**Comment by davidrapin** [comment]
@icewind1991 @jancborchardt Anything else missing for this to be merged?

**Comment by icewind1991** [comment]
:+1:

**Comment by jancborchardt** [comment]
The only thing I’m hesitant about is that this adds a JS library of about 10.000 lines of code. Opinions @PVince81 @DeepDiver1975?

**Comment by DeepDiver1975** [comment]
> The only thing I’m hesitant about is that this adds a JS library of about 10.000 lines of code. Opinions @PVince81 @DeepDiver1975?

That lib contains a lot of comments (which is great) - in a deployment environment where the asset pipline is enabled this will have no impact from my understanding. :+1:

**Comment by DeepDiver1975** [comment]
needs rebase

**Comment by davidrapin** [comment]
@DeepDiver1975 rebase done :)

**Comment by oparoz** [comment]
:+1: 
Added to the Gallery rewrite. It's a really cool feature to have. Thanks @davidrapin .

**Comment by davidrapin** [comment]
@oparoz You're welcome.

**Comment by oparoz** [comment]
I've made the following changes
- Images are initially shown at their native size if they're smaller than the screen area
- Small images are initially magnified so that they don't appear too tiny. They can also be zoomed more
- '9' zooms to fit
- '0' zooms to original size
- Images zoom to fit when going fullscreen. No point in having small images there

You can see what it looks like in the [demo](https://oc7demo.interfacloud.com/index.php/apps/galleryplus/public?token=312b038d3e70877199a4d66abf37bdd0) and the [source code for OC7](https://github.com/interfasys/galleryplus/tree/stable7) is now finally available.

**Comment by davidrapin** [comment]
@oparoz nice !

**Comment by oparoz** [comment]
Let's close this as it's already been implemented in Gallery.

**Pull request closed without merging**